velox/functions/lib/DateTimeFormatter.cpp

```diff
@@ -15,9 +15,281 @@
+  auto contentLength = strContent.size();
+  if (contentLength == 0) {
+    return std::string(totalDigits, padding);
+  }
+
+  bool isNegative = strContent[0] == '-';
+  auto digitLength = contentLength - (isNegative ? 1 : 0);
+  if (digitLength >= totalDigits) {
+    return strContent;
+  }
+  std::string paddingStr(totalDigits - digitLength, padding);
+  if (padFront) {
+    return strContent.insert((isNegative ? 1 : 0), paddingStr);
+  } else {
+    return strContent.append(paddingStr);
+  }
+}
+
+template <>
+std::string padContent<std::string>(
+    const std::string& content,
+    char padding,
+    size_t totalDigits,
+    bool padFront) {
+  auto contentLength = content.size();
+  if (contentLength >= totalDigits) {
+    return content;
+  }
+  std::string paddingStr(totalDigits - contentLength, padding);
+  if (padFront) {
+    return paddingStr + content;
+  } else {
+    return content + paddingStr;
+  }
+}
+
+void validateTimePoint(const std::chrono::time_point<
+                       std::chrono::system_clock,
+                       std::chrono::milliseconds>& timePoint) {
+  // Due to the limit of std::chrono we can only represent time in
+  // [-32767-01-01, 32767-12-31] date range
+  const auto minTimePoint = date::sys_days{
+      date::year_month_day(date::year::min(), date::month(1), date::day(1))};
+  const auto maxTimePoint = date::sys_days{
+      date::year_month_day(date::year::max(), date::month(12), date::day(31))};
+  if (timePoint < minTimePoint || timePoint > maxTimePoint) {
+    VELOX_USER_FAIL(
+        "Cannot format time out of range of [{}-{}-{}, {}-{}-{}]",
+        (int)date::year::min(),
+        "01",
+        "01",
+        (int)date::year::max(),
+        "12",
+        "31");
+  }
+}
+
+} // namespace
+
+std::string DateTimeFormatter::format(
+    const Timestamp& inputTs,
+    const date::time_zone& timezone) const {
+  Timestamp timestamp = inputTs;
+  timestamp.toTimezone(timezone);
+  const std::chrono::
+      time_point<std::chrono::system_clock, std::chrono::milliseconds>
+          timePoint(std::chrono::milliseconds(timestamp.toMillis()));
+  validateTimePoint(timePoint);
+  const auto daysTimePoint = date::floor<date::days>(timePoint);
+
+  const auto durationInTheDay = date::make_time(timePoint - daysTimePoint);
+  const date::year_month_day calDate(daysTimePoint);
+  const date::weekday weekday(daysTimePoint);
+  if (!calDate.ok()) {
+    VELOX_USER_FAIL("Invalid date");
+  }
+  std::string result;
+  // TODO: check if we have timezone in it. If we do, we always format to local
+  // timezone
+  for (auto& token : tokens_) {
+    if (token.type == DateTimeToken::Type::kLiteral) {
+      result += token.literal;
+    } else {
+      switch (token.pattern.specifier) {
+        case DateTimeFormatSpecifier::ERA:
+          result += static_cast<signed>(calDate.year()) >= 0 ? "AD" : "BC";
+          break;
+        case DateTimeFormatSpecifier::CENTURY_OF_ERA: {
```

Comment:
I wonder if for readability it is better to leave empty lines between cases?
Otherwise hard to see where the next case starts.